Idempotent in headless mode

jmeubank opened the pull request:
This changeset adjusts the script so that you can run it multiple times with the same input and not have any unexpected changes. This makes it appropriate for "enforcing state", as required by automated provisioners like Puppet, Salt, Chef, or Ansible.

 - Unbound, OpenVPN, easy-rsa, and other dependencies are only installed from upstream if they are not already present. This prevents multiple runs of the script from causing unexpected version upgrades.
 - The easy-rsa CA is only initialized once
 - SERVER_CN and SERVER_NAME are randomly generated once and saved for future reference
 - File append ('>>') is only done strictly after a file is created with '>' (e.g. /etc/sysctl.d/20-openvpn.conf)
 - Clients are only added to easy-rsa once
 - If AUTO_INSTALL == y, then the script operates in install mode and doesn't enter manageMenu

Also, IPv4-only mode has been fixed with a call to curl -4 ifconfig.co.

Fixes: #545

jmeubank committed:
Idempotent in headless mode

This set of changes adjusts the script so that you can run it multiple times with the same input and not have any unexpected changes. This makes it appropriate for "enforcing state", as required by automated provisioners like Puppet, Salt, Chef, or Ansible.

 - Unbound, OpenVPN, easy-rsa, and other dependencies are only installed from upstream if they are not already present. This prevents multiple runs of the script from causing unexpected version upgrades.
 - The easy-rsa system is put in a folder called "easy-rsa-auto" so it can't conflict with the "easy-rsa" folder from some older OpenVPN packages
 - The easy-rsa CA is only initialized once
 - SERVER_CN and SERVER_NAME are randomly generated once and saved for future reference
 - File append ('>>') is only done strictly after a file is created with '>' (e.g. /etc/sysctl.d/20-openvpn.conf)
 - Clients are only added to easy-rsa once
 - If AUTO_INSTALL == y, then the script operates in install mode and doesn't enter manageMenu

randshell commented on openvpn-install.sh:
```diff
@@ -596,8 +599,12 @@ function installOpenVPN () {
+		else
+			PUBLIC_IPV4=$(curl -4 ifconfig.co)
+		fi
 		ENDPOINT=${ENDPOINT:-$PUBLIC_IPV4}
```
@angristan are we sure we can change this behavior to default to an IPv6 endpoint?

angristan replied:
I think most people would still prefer to connect to an IPv4 endpoint...

HenryNe replied:
I prefer also an IPv4 endpoint.
An IPv6 endpoint should only a fallback, if IPv4 is not usable.

For my understand the IPV6_SUPPORT is a define to support IPv6 and NAT inside the VPN.
For a dual stack endpoint I enable IPV6_SUPPORT and put a FDNS name as endpoint.

jmeubank replied:
curl ifconfig.co in the current version returns an IPv6 address on my servers, which is why I made this change - although my servers are IPv6 enabled, I'm disabling IPv6 in OpenVPN (and forcing an IPv6 blackhole with route-ipv6 2000::/3 dead::beef). I could set ENDPOINT separately from the script, so this isn't a make-or-break change for me, but it was an easy fix. And for dynamically assigned IPs, it's nice to have.

randshell replied:
@jmeubank Can you please change it to `https://ifconfig.co` to use the https version? You are already working on this line so I'm not going to open a new PR.

jmeubank replied:
No prob - done.

jmeubank committed:
Mention Ansible in README under idempotent headless mode

randshell commented on openvpn-install.sh:
```diff
@@ -623,33 +630,38 @@ function installOpenVPN () {
-	elif [[ "$OS" = 'arch' ]]; then
-		# Install required dependencies and upgrade the system
-		pacman --needed --noconfirm -Syu openvpn iptables openssl wget ca-certificates curl
 	fi
```
```suggestion
		if [[ -d /etc/openvpn/easy-rsa/ ]]; then
			rm -rf /etc/openvpn/easy-rsa/
		fi
	fi
```

randshell commented on openvpn-install.sh:
```diff
@@ -664,60 +676,72 @@ function installOpenVPN () {
-	rm -f ~/EasyRSA-unix-v${version}.tgz
+	# Install the latest version of easy-rsa from source, if not already
+	# installed.
+	if [[ ! -d /etc/openvpn/easy-rsa-auto/ ]]; then
```
```suggestion
	if [[ ! -d /etc/openvpn/easy-rsa/ ]]; then
```

randshell commented on openvpn-install.sh:
```diff
@@ -664,60 +676,72 @@ function installOpenVPN () {
+		tar xzf ~/EasyRSA-unix-v${version}.tgz -C ~/
+		mkdir -p /etc/openvpn/easy-rsa-auto
+		mv ~/EasyRSA-v${version}/* /etc/openvpn/easy-rsa-auto/
+		chown -R root:root /etc/openvpn/easy-rsa-auto/
```
```suggestion
		mkdir -p /etc/openvpn/easy-rsa
		mv ~/EasyRSA-v${version}/* /etc/openvpn/easy-rsa/
		chown -R root:root /etc/openvpn/easy-rsa/
```

randshell commented on openvpn-install.sh:
```diff
@@ -664,60 +676,72 @@ function installOpenVPN () {
+		chown -R root:root /etc/openvpn/easy-rsa-auto/
+		rm -f ~/EasyRSA-unix-v${version}.tgz
+
+		cd /etc/openvpn/easy-rsa-auto/ || return
```
```suggestion
		cd /etc/openvpn/easy-rsa/ || return
```

randshell commented on openvpn-install.sh:
```diff
@@ -664,60 +676,72 @@ function installOpenVPN () {
+	else
+		# If easy-rsa is already installed, grab the generated SERVER_NAME
+		# for client configs
+		cd /etc/openvpn/easy-rsa-auto/ || return
```
```suggestion
		cd /etc/openvpn/easy-rsa/ || return
```

randshell commented on openvpn-install.sh:
```diff
@@ -664,60 +676,72 @@ function installOpenVPN () {
 
 	# Move all the generated files
-	cp pki/ca.crt pki/private/ca.key "pki/issued/$SERVER_NAME.crt" "pki/private/$SERVER_NAME.key" /etc/openvpn/easy-rsa/pki/crl.pem /etc/openvpn
+	cp pki/ca.crt pki/private/ca.key "pki/issued/$SERVER_NAME.crt" "pki/private/$SERVER_NAME.key" /etc/openvpn/easy-rsa-auto/pki/crl.pem /etc/openvpn
```
```suggestion
	cp pki/ca.crt pki/private/ca.key "pki/issued/$SERVER_NAME.crt" "pki/private/$SERVER_NAME.key" /etc/openvpn/easy-rsa/pki/crl.pem /etc/openvpn
```

randshell commented on openvpn-install.sh:
```diff
@@ -1026,16 +1050,23 @@ function newClient () {
-			./easyrsa build-client-full "$CLIENT"
-		;;
-	esac
+	CLIENTEXISTS=$(tail -n +2 /etc/openvpn/easy-rsa-auto/pki/index.txt | grep -c -E "/CN=$CLIENT\$")
```
```suggestion
	CLIENTEXISTS=$(tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep -c -E "/CN=$CLIENT\$")
```

randshell commented on openvpn-install.sh:
```diff
@@ -1026,16 +1050,23 @@ function newClient () {
+		echo ""
+		echo "The specified client CN was found in easy-rsa."
+	else
+		cd /etc/openvpn/easy-rsa-auto/ || return
```
```suggestion
		cd /etc/openvpn/easy-rsa/ || return
```

randshell commented on openvpn-install.sh:
```diff
@@ -1057,15 +1088,15 @@ function newClient () {
 	{
 		echo "<ca>"
-		cat "/etc/openvpn/easy-rsa/pki/ca.crt"
+		cat "/etc/openvpn/easy-rsa-auto/pki/ca.crt"
```
```suggestion
		cat "/etc/openvpn/easy-rsa/pki/ca.crt"
```

randshell commented on openvpn-install.sh:
```diff
@@ -1057,15 +1088,15 @@ function newClient () {
 
 		echo "<cert>"
-		awk '/BEGIN/,/END/' "/etc/openvpn/easy-rsa/pki/issued/$CLIENT.crt"
+		awk '/BEGIN/,/END/' "/etc/openvpn/easy-rsa-auto/pki/issued/$CLIENT.crt"
```
```suggestion
		awk '/BEGIN/,/END/' "/etc/openvpn/easy-rsa/pki/issued/$CLIENT.crt"
```

randshell commented on openvpn-install.sh:
```diff
@@ -1057,15 +1088,15 @@ function newClient () {
 
 		echo "<key>"
-		cat "/etc/openvpn/easy-rsa/pki/private/$CLIENT.key"
+		cat "/etc/openvpn/easy-rsa-auto/pki/private/$CLIENT.key"
```
```suggestion
		cat "/etc/openvpn/easy-rsa/pki/private/$CLIENT.key"
```

randshell commented on openvpn-install.sh:
```diff
@@ -1084,14 +1115,14 @@ function newClient () {
 
 function revokeClient () {
-	NUMBEROFCLIENTS=$(tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep -c "^V")
+	NUMBEROFCLIENTS=$(tail -n +2 /etc/openvpn/easy-rsa-auto/pki/index.txt | grep -c "^V")
```
```suggestion
	NUMBEROFCLIENTS=$(tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep -c "^V")
```

randshell commented on openvpn-install.sh:
```diff
@@ -1100,23 +1131,23 @@ function revokeClient () {
 	echo ""
 	echo "Select the existing client certificate you want to revoke"
-	tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep "^V" | cut -d '=' -f 2 | nl -s ') '
+	tail -n +2 /etc/openvpn/easy-rsa-auto/pki/index.txt | grep "^V" | cut -d '=' -f 2 | nl -s ') '
```
```suggestion
	tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep "^V" | cut -d '=' -f 2 | nl -s ') '
```

randshell commented on openvpn-install.sh:
```diff
@@ -1100,23 +1131,23 @@ function revokeClient () {
-	CLIENT=$(tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep "^V" | cut -d '=' -f 2 | sed -n "$CLIENTNUMBER"p)
-	cd /etc/openvpn/easy-rsa/ || return
+	CLIENT=$(tail -n +2 /etc/openvpn/easy-rsa-auto/pki/index.txt | grep "^V" | cut -d '=' -f 2 | sed -n "$CLIENTNUMBER"p)
+	cd /etc/openvpn/easy-rsa-auto/ || return
```
```suggestion
	CLIENT=$(tail -n +2 /etc/openvpn/easy-rsa/pki/index.txt | grep "^V" | cut -d '=' -f 2 | sed -n "$CLIENTNUMBER"p)
	cd /etc/openvpn/easy-rsa/ || return
```

randshell commented on openvpn-install.sh:
```diff
@@ -1100,23 +1131,23 @@ function revokeClient () {
 	rm -f "pki/issued/$CLIENT.crt"
 	rm -f /etc/openvpn/crl.pem
-	cp /etc/openvpn/easy-rsa/pki/crl.pem /etc/openvpn/crl.pem
+	cp /etc/openvpn/easy-rsa-auto/pki/crl.pem /etc/openvpn/crl.pem
```
```suggestion
	cp /etc/openvpn/easy-rsa/pki/crl.pem /etc/openvpn/crl.pem
```

randshell reviewed (changes requested):
> subsequent runs of this script shouldn't re-download and recreate it.

Now I get it. It would happen when you try to run the command to install it again but it's already installed.

I think we can avoid changing the name of the `easy-rsa` directory to `easy-rsa-auto` if we move `rm -rf /etc/openvpn/easy-rsa/` to be inside `if [[ ! -e /etc/openvpn/server.conf ]]; then`. Can you let me know if you agree please?

jmeubank committed:
Use original easy-rsa name instead of easy-rsa-auto

Co-Authored-By: randomshell <43271778+randomshell@users.noreply.github.com>

jmeubank committed:
Use original easy-rsa name (round 2)

Co-Authored-By: randomshell <43271778+randomshell@users.noreply.github.com>

randshell reviewed (approved):
Thanks!

jmeubank committed:
Use HTTPS when determining public IPv4/IPv6

angristan commented on openvpn-install.sh:
```diff
@@ -596,8 +599,12 @@ function installOpenVPN () {
+		else
+			PUBLIC_IPV4=$(curl -4 https://ifconfig.co)
+		fi
 		ENDPOINT=${ENDPOINT:-$PUBLIC_IPV4}
```
I guess we can rename `PUBLIC_IPV4` to `PUBLIC_IP` now 🙂

jmeubank replied:
Easy, done. :)

jmeubank committed:
Change PUBLIC_IPV4 to PUBLIC_IP for IPv4/v6

angristan commented on openvpn-install.sh:
```diff
@@ -664,57 +679,69 @@ function installOpenVPN () {
-	rm -f ~/EasyRSA-unix-v${version}.tgz
+	# Install the latest version of easy-rsa from source, if not already
+	# installed.
+	if [[ ! -d /etc/openvpn/easy-rsa/ ]]; then
```
This is ineffective because we delete the folder right before :)

randshell replied:
I suggested to @jmeubank to move the deletion code of easy-rsa that comes with openvpn to be after the openvpn installation to avoid renaming to easy-rsa-auto https://github.com/angristan/openvpn-install/pull/592#pullrequestreview-383455081

@angristan You are right, I missed the fact that we didn't delete the old remove code so now easy-rsa gets removed every time anyway.
https://github.com/angristan/openvpn-install/blob/5805cc03b2a45fb9dd859b60a5e6fe30d2cf28a2/openvpn-install.sh#L677-L680
This should be removed.

jmeubank replied:
Yeah. I think I screwed that up with my rebase. Fixing!

randshell reviewed (changes requested):
Thanks @angristan

jmeubank committed:
Fix easy-rsa old version removal

The removal of an old version of easy-rsa should only happen if OpenVPN is being installed for the first time.
